### PR TITLE
Outcome Service Refactor

### PR DESCRIPTION
**Pull request by JacobHarrisonQA**
cleaned up the logic used within templating so it calls classes based on outcomes types

### CODE REVIEW

File: src/main/java/uk/gov/ons/census/fwmt/outcomeservice/converter/interview/InterviewAddressNotValidProcessor.java [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+      gatewayOutcomeProducer.sendAddressUpdate(outcomeEvent, String.valueOf(ccsInterviewOutcome.getTransactionId()));
+      gatewayEventManager.triggerEvent(String.valueOf(ccsInterviewOutcome.getCaseId()), OUTCOME_SENT_RM, LocalTime.now());
+    } catch (GatewayException e) {
+      e.printStackTrace();
+    }
+  }
+
+  private boolean isNonValid(CCSInterviewOutcome ccsInterviewOutcome) {
+    List<String> validSecondaryOutcomes = Arrays
+        .asList("Derelict", "Demolished", "Property is a CE", "Unaddressable Object", "Non residential or business",
```
Comment (spdiaz1000) [inline src/main/java/uk/gov/ons/census/fwmt/outcomeservice/converter/interview/InterviewAddressNotValidProcessor.java]: I dont see reference to Unaddressable Object in the confluence doc

File: src/main/java/uk/gov/ons/census/fwmt/outcomeservice/converter/interview/InterviewAddressNotValidProcessor.java [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+    root.put("secondaryOutcome",
+        InterviewSecondaryOutcomeMap.interviewSecondaryOutcomeMap.get(ccsInterviewOutcome.getSecondaryOutcome()));
+
+    String outcomeEvent = TemplateCreator.createOutcomeMessage(ADDRESS_NOT_VALID, root, household);
+
+    try {
+      gatewayOutcomeProducer.sendAddressUpdate(outcomeEvent, String.valueOf(ccsInterviewOutcome.getTransactionId()));
+      gatewayEventManager.triggerEvent(String.valueOf(ccsInterviewOutcome.getCaseId()), OUTCOME_SENT_RM, LocalTime.now());
+    } catch (GatewayException e) {
+      e.printStackTrace();
```
Comment (spdiaz1000) [inline src/main/java/uk/gov/ons/census/fwmt/outcomeservice/converter/interview/InterviewAddressNotValidProcessor.java]: See comment on AddressNotValidProcessor.java

File: src/main/java/uk/gov/ons/census/fwmt/outcomeservice/converter/interview/InterviewQuestionnaireLinkedProcessor.java [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+      root.put("ccsInterviewOutcome", ccsInterviewOutcome);
+      root.put("questionnaireId", ccsInterviewOutcome.getFulfillmentRequest().getQuestionnaireId());
+
+      String outcomeEvent = TemplateCreator.createOutcomeMessage(QUESTIONNAIRE_LINKED, root, interview);
+
+      try {
+        gatewayOutcomeProducer.sendFulfilmentRequest(outcomeEvent, String.valueOf(ccsInterviewOutcome.getTransactionId()));
+        gatewayEventManager.triggerEvent(String.valueOf(ccsInterviewOutcome.getCaseId()), OUTCOME_SENT_RM, LocalTime.now());
+      } catch (GatewayException e) {
+        e.printStackTrace();
```
Comment (spdiaz1000) [inline src/main/java/uk/gov/ons/census/fwmt/outcomeservice/converter/interview/InterviewQuestionnaireLinkedProcessor.java]: See comment on AddressNotValidProcessor.java

File: src/main/java/uk/gov/ons/census/fwmt/outcomeservice/converter/interview/InterviewQuestionnaireLinkedProcessor.java [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+  }
+
+  private boolean isContactMade(CCSInterviewOutcome ccsInterviewOutcome) {
+    List<String> validCompletion = Collections.singletonList("Left questionnaire on final visit");
+    return ccsInterviewOutcome.getPrimaryOutcome().equals(CONTACT_MADE.toString()) && validCompletion
+        .contains(ccsInterviewOutcome.getSecondaryOutcome());
+  }
+
+  private boolean isNoContact(CCSInterviewOutcome ccsInterviewOutcome) {
+    List<String> validCompletion = Arrays.asList("Complete on paper (full)", "Complete on paper (partial)",
```
Comment (spdiaz1000) [inline src/main/java/uk/gov/ons/census/fwmt/outcomeservice/converter/interview/InterviewQuestionnaireLinkedProcessor.java]: I only see Contact Made with "Complete on paper (full)", "Complete on paper (partial)" in confluence

File: src/main/java/uk/gov/ons/census/fwmt/outcomeservice/converter/interview/InterviewRefusalReceivedProcessor.java [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+    root.put("refusalType",
+        InterviewSecondaryOutcomeMap.interviewSecondaryOutcomeMap.get(ccsInterviewOutcome.getSecondaryOutcome()));
+
+    String outcomeEvent = TemplateCreator.createOutcomeMessage(REFUSAL_RECEIVED, root, interview);
+
+    try {
+      gatewayOutcomeProducer.sendRespondentRefusal(outcomeEvent, String.valueOf(ccsInterviewOutcome.getTransactionId()));
+      gatewayEventManager.triggerEvent(String.valueOf(ccsInterviewOutcome.getCaseId()), OUTCOME_SENT_RM, LocalTime.now());
+    } catch (GatewayException e) {
+      e.printStackTrace();
```
Comment (spdiaz1000) [inline src/main/java/uk/gov/ons/census/fwmt/outcomeservice/converter/interview/InterviewRefusalReceivedProcessor.java]: See comment on AddressNotValidProcessor.java

File: src/main/java/uk/gov/ons/census/fwmt/outcomeservice/converter/interview/InterviewRefusalReceivedProcessor.java [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+    }
+  }
```
Comment (spdiaz1000) [inline src/main/java/uk/gov/ons/census/fwmt/outcomeservice/converter/interview/InterviewRefusalReceivedProcessor.java]: the ones missing are
Contact Made with "Complete on paper (full)", "Complete on paper (partial)"

File: src/main/java/uk/gov/ons/census/fwmt/outcomeservice/converter/interview/InterviewSecondaryOutcomeMap.java [hunk 7]
```diff
@@ -0,0 +1,28 @@
+package uk.gov.ons.census.fwmt.outcomeservice.converter.interview;
+
+import java.util.HashMap;
+import java.util.Map;
+
+public class InterviewSecondaryOutcomeMap {
+
```
Comment (spdiaz1000) [inline src/main/java/uk/gov/ons/census/fwmt/outcomeservice/converter/interview/InterviewSecondaryOutcomeMap.java]: create a private default constructor and get other classes to connect in a static way

File: src/test/java/uk/gov/ons/census/fwmt/outcomeservice/helper/HouseholdOutcomeBuilder.java [hunk 8]
```diff
@@ -392,83 +387,83 @@ public static HouseholdOutcome createCENoValidAddressNoContactMade() {
...
-//      
-//      HouseholdOutcome householdOutcome18 = createCENoValidAddressNoContactMade();
-//      ho = mapper.writerWithDefaultPrettyPrinter().writeValueAsString(householdOutcome18);
-//      System.out.println(ho);
-//      
-//  }
-  
+
+  //  public static void main(String[] args) throws JsonGenerationException, JsonMappingException, IOException {
+  //    ObjectMapper mapper = new ObjectMapper();
```
Comment (spdiaz1000) [inline src/test/java/uk/gov/ons/census/fwmt/outcomeservice/helper/HouseholdOutcomeBuilder.java]: whats all this then

File: src/test/java/uk/gov/ons/census/fwmt/outcomeservice/message/GatewayOutcomeProducerTest.java [hunk 9]
```diff
@@ -1,119 +1,100 @@
...
+  //  ObjectMapper objectMapper;
 
-//  @Test
-//  public void sendAddressUpdateTest() throws JsonProcessingException, GatewayException {
-//    //Given
-//    OutcomeEvent outcomeEvent = new OutcomeEvent();
-//    Event event = new Event();
-//    event.setTransactionId(UUID.fromString("c45de4dc-3c3b-11e9-b210-d663bd873d93"));
-//    outcomeEvent.setEvent(event);
-//
```
Comment (spdiaz1000) [inline src/test/java/uk/gov/ons/census/fwmt/outcomeservice/message/GatewayOutcomeProducerTest.java]: whats all this then

File: src/main/java/uk/gov/ons/census/fwmt/outcomeservice/converter/interview/InterviewSecondaryOutcomeMap.java (hunk 7 above)
Comment (JacobHarrisonQA) [reply src/main/java/uk/gov/ons/census/fwmt/outcomeservice/converter/interview/InterviewSecondaryOutcomeMap.java]: Still having issues with getting this to work as intended

File: src/main/java/uk/gov/ons/census/fwmt/outcomeservice/converter/ccs/CcsRefusalReceivedProcessor.java [hunk 10]
```diff
@@ -0,0 +1,60 @@
...
+  @Autowired
+  private GatewayOutcomeProducer gatewayOutcomeProducer;
+
+  @Autowired
+  private GatewayEventManager gatewayEventManager;
+
+  @Override
+  public boolean isValid(CCSPropertyListingOutcome ccsPropertyListingOutcome) {
+    List<String> validSecondaryOutcomes = Arrays
+        .asList("Soft refusal", "Hard refusal", "Extraordinary refusal");
```
Comment (garethTurner) [inline src/main/java/uk/gov/ons/census/fwmt/outcomeservice/converter/ccs/CcsRefusalReceivedProcessor.java]: not worth changing but you should be able to just use List.of()

**Pull request merged**